transport.go

```diff
@@ -43,37 +40,28 @@ func (t *transport) RoundTrip(ctx *fasthttp.RequestCtx) {
 			ctx.Error("", fasthttp.StatusServiceUnavailable)
 			return
 		}
-		t.mu.Unlock()
-		if bytes.HasSuffix(umaeEndpoint, ctx.Path()) {
-			ctx.Success("text/plain", t.getUMAE())
-		}
-
-		finished := t.waiter.finishedCount()
-		if finished > 0 && finished%t.window.size() == 0 {
-			ctx.Error("", fasthttp.StatusServiceUnavailable)
-			go t.checkHeapAndGC()
-			return
-		}
 		t.waiter.requestArrived()
+		t.mu.Unlock()
 	}
 	ctx.Request.Header.Del("Connection")
 	if err := t.client.Do(&ctx.Request, &ctx.Response); err != nil {
 		panic(fmt.Sprintf("Problem calling proxy:%q", err))
 	}
 	ctx.Response.Header.Del("Connection")
 	if !*disableGCI {
-		t.mu.Lock()
-		t.waiter.requestFinished()
-		t.mu.Unlock()
+		if t.waiter.requestFinished()%t.window.size() == 0 {
+			// Yes, it is possible to call checkHeapAndGC twice
+			// before the service becomes unavailable.
+			t.mu.Lock()
+			if !t.checking {
+				t.checking = true
```

Comment:
Shouldn't that be set inside checkHeapAndGC method?

Reply:
Wouldn't be straightforward just call checkHeapAndGC and let that method verify and update t.checking?

Reply:
> Wouldn't be straightforward just call checkHeapAndGC and let that method verify and update t.checking?

No because we can trigger the go routine twice. Notice that setting the variable to true and triggering the go routine is done atomically.

Reply:
> > Wouldn't be straightforward just call checkHeapAndGC and let that method verify and update t.checking?
> 
> No because we can trigger the go routine twice. Notice that setting the variable to true and triggering the go routine is done atomically.

If the checking were done inside that method, it wouldn't impact the program accuracy, I believe.  I confess that way (my suggestion) isn't too efficiency, 'cause a new goroutine is created only to check something...